{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "AMKZ5QVoBmjJ"
   },
   "source": [
    "<h1><b>Statistique en Bioinformatique : </b> TME9 </h1><br>\n",
    "\n",
    "L’objectif de ce TME est l'implementation de la méthode Expectation-Maximisation pour la recherche de motifs.\n",
    "\n",
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"margin: 10px\">\n",
    "<div class=\"alert alert-warning\" role=\"alert\" style=\"margin: 10px\">\n",
    "<p><b>Soumission</b></p>\n",
    "<ul>\n",
    "<li>Renomer le fichier TME9.ipynb pour NomEtudiant1_NomEtudiant2.ipynb </li>\n",
    "<li>Soumettre via moodle </li>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "r_qH1CG7BmjM"
   },
   "source": [
    "<H1>Expectation-Maximisation Motif</H1>\n",
    "<br>\n",
    "La méthode EM (Expectation-Maximisation) permet de détecter des motifs dans un ensemble de séquences ADN ou protéiques reliées, non alignées. En particulier, étant donné un groupe de séquences de longueur $L$, dont on sait qu'elles partagent un motif commun de longueur $w$, l’algorithme EM:\n",
    "- infère un modèle $(\\Theta,Z)$ pour le motif;\n",
    "- localise l’occurrence du motif dans chaque séquence.\n",
    "\n",
    "$\\Theta$ representé la matrice des poids-positions $p_{c,k}$ du motif (où $c \\in \\{A,C,G,T\\}$ ou $c \\in \\{A,C,D,...,W\\}$  et $k \\in \\{0 \\dots w\\}$), $p_{c,0}$  est le vecteur de probabilités du modèle nul ou \"background\".\n",
    "$Ζ$ est la matrice des variables cachées, qui donnent les positions initiales du motif: \n",
    "- $Z_{i,j} = 1$, si le motif commence en position $j$ de la séquence $i$,\n",
    "- $Z_{i,j} = 0$, sinon. \n",
    "\n",
    "L’algorithme affine les paramètres du modèle de manière itérative par espérance-maximisation. Chaque itération $t$ se compose de deux étapes:\n",
    "- (E) Calcul des valeurs attendues $Ζ^{(t)}$ de $Ζ$, étant donnés $\\Theta^{(t-1)}$\n",
    "- (M) Estimation de  $\\Theta^{(t)}$  à partir de  $Ζ^{(t)}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "uSue1iYeBmjO"
   },
   "source": [
    "1\\. Implémentez une fonction `read_training_file` pour lire le fichier d'entré qui contient un ensemble de séquences ADN non alignées. Pour simplifier nous allons utiliser les données vu en cours du fichier `toyEx.txt`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "id": "aWQrgN0iBmjP"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GTCAGG', 'GAGAGT', 'ACGGAG', 'CCAGTC']\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "nts = ['A', 'C', 'G', 'T']\n",
    "\n",
    "w = 3\n",
    "input_f = \"toyEx.txt\"\n",
    "\n",
    "def read_training_file(input_f):\n",
    "    \"\"\"\n",
    "    Read a file with no-aligned sequences\n",
    "    input input_f : file name\n",
    "    output seqs : list of sequences\n",
    "    \"\"\"\n",
    "    seqs = []\n",
    "    \n",
    "    f = open(input_f)\n",
    "    for lines in f.readlines():\n",
    "        seqs.append(lines[:-1])\n",
    "   \n",
    "    return seqs\n",
    "\n",
    "seqs = read_training_file(input_f)\n",
    "print (seqs) #['GTCAGG', 'GAGAGT', 'ACGGAG', 'CCAGTC']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XIHpbbKwBmjQ"
   },
   "source": [
    "2\\. Implémentez une fonction `initialiseP` pour initialiser la matrice poids-position $p_{c,k}$. On considère le modèle nul par défaut $p_0 = (0.25, 0.25, 0.25, 0.25)$. Pour initialiser $p^{(t)}$, on prend généralement un motif au hasard dans une sequence, et on fixe à $0.5$ les poids du nucleotide correspondant et à $\\frac{1-0.5}{3}$ les trois autres. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "id": "_zIZdJouBmjR"
   },
   "outputs": [],
   "source": [
    "# def initialiseP(seqs, w, alph):\n",
    "#     \"\"\"\n",
    "#     Initialise pc,k\n",
    "#     input seqs : list of sequences\n",
    "#     input w : motif length\n",
    "#     input alph : alphabet (nucleotides or amino acids)\n",
    "#     output P: position probability matrix\n",
    "#     \"\"\"\n",
    "#     q=len(alph)\n",
    "#     P = np.zeros((q, w+1))\n",
    "    \n",
    "#     for j in range (q): \n",
    "#         P[j][0]= 0.25\n",
    "        \n",
    "#     print(w)\n",
    "    \n",
    "#     seqr=np.random.randint(len(seqs))\n",
    "#     SEQ = seqs[seqr]\n",
    "#     print(\"SEQ\", SEQ)\n",
    "#     r =np.random.randint(0,(len(SEQ)-w))\n",
    "#     motif = SEQ[r:(r+w)]\n",
    "#     print(\"motif\",motif)\n",
    "    \n",
    "#     for i in range(1,len(motif)):\n",
    "#         print(\"i\", i)\n",
    "#         indice = alph.index(motif[i])\n",
    "#         print(\"indice\", indice)\n",
    "#         P[indice][i] = 0.5\n",
    "        \n",
    "#         for j in  range (len(motif)+1): \n",
    "#             if j != indice: \n",
    "#                 P[j][i] = (1 - 0.5)/3\n",
    "# #         if motif[i]==alpha\n",
    "    \n",
    "    \n",
    "    \n",
    "#     lnt=[]   \n",
    "# #     for i in range(len(seqs)):\n",
    "    \n",
    "#     r=np.random.randint(0,len(seqs[0]))\n",
    "#     nt=seqs[0][r]\n",
    "# #         lnt.append(nt)\n",
    "#     print(nt)\n",
    "        \n",
    "# #         if nt=='A':\n",
    "# #             ind=0\n",
    "# #         if nt=='C':\n",
    "# #             ind=1\n",
    "# #         if nt=='G':\n",
    "# #             ind=2\n",
    "# #         if nt=='T':\n",
    "# #             ind=3\n",
    "        \n",
    "       \n",
    "# #         P[ind][ind]=0.5\n",
    "    \n",
    "    \n",
    "    \n",
    "#     #alph[r]=0.5\n",
    "    \n",
    "   \n",
    "   \n",
    "#     return P\n",
    "\n",
    "# #test\n",
    "# p = initialiseP(seqs, w, nts)\n",
    "# print (p)\n",
    "# print(nts)\n",
    "# print(np.random.randint(0,4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3\n",
      "SEQ CCAGTC\n",
      "motif AGT\n",
      "MOTIF[i] A\n",
      "indice dans boucle 0\n",
      "MOTIF[i] G\n",
      "indice dans boucle 2\n",
      "MOTIF[i] T\n",
      "indice dans boucle 3\n",
      "T\n",
      "[[0.25       0.5        0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.5        0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.5       ]]\n",
      "['A', 'C', 'G', 'T']\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "def initialiseP(seqs, w, alph):\n",
    "    \"\"\"\n",
    "    Initialise pc,k\n",
    "    input seqs : list of sequences\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output P: position probability matrix\n",
    "    \"\"\"\n",
    "    q=len(alph)\n",
    "    P = np.zeros((q, w+1))\n",
    "    \n",
    "    for j in range (q): \n",
    "        P[j][0]= 0.25\n",
    "        \n",
    "    print(w)\n",
    "    \n",
    "    seqr=np.random.randint(len(seqs))\n",
    "    SEQ = seqs[seqr]\n",
    "    print(\"SEQ\", SEQ)\n",
    "    r =np.random.randint(0,(len(SEQ)-w))\n",
    "    motif = SEQ[r:(r+w)]\n",
    "    print(\"motif\",motif)\n",
    "    \n",
    "    \n",
    "    lind=[]\n",
    "    for i in range(len(motif)):\n",
    "        print(\"MOTIF[i]\",motif[i])\n",
    "        indice = alph.index(motif[i])\n",
    "        \n",
    "        print(\"indice dans boucle\",indice)\n",
    "        lind.append(indice)\n",
    "        \n",
    "    for j in range(len(lind)):\n",
    "           \n",
    "        P[lind[j]][j+1]=0.5\n",
    "    \n",
    "        \n",
    "    for i in range(1,w+1):\n",
    "        for j in range(q):\n",
    "            if P[j][i]==0.5:\n",
    "                continue\n",
    "            else:\n",
    "                P[j][i]=((1-0.5)/3)\n",
    "        \n",
    "#         for j in  range (len(motif)): \n",
    "#             if j != indice: \n",
    "#                 P[j][i] = (1 - 0.5)/3\n",
    "# #         if motif[i]==alpha\n",
    "    \n",
    "    \n",
    "    \n",
    "    lnt=[]   \n",
    "#     for i in range(len(seqs)):\n",
    "    \n",
    "    r=np.random.randint(0,len(seqs[0]))\n",
    "    nt=seqs[0][r]\n",
    "#         lnt.append(nt)\n",
    "    print(nt)\n",
    "        \n",
    "#         if nt=='A':\n",
    "#             ind=0\n",
    "#         if nt=='C':\n",
    "#             ind=1\n",
    "#         if nt=='G':\n",
    "#             ind=2\n",
    "#         if nt=='T':\n",
    "#             ind=3\n",
    "        \n",
    "       \n",
    "#         P[ind][ind]=0.5\n",
    "    \n",
    "    \n",
    "    \n",
    "    #alph[r]=0.5\n",
    "    \n",
    "   \n",
    "   \n",
    "    return P\n",
    "\n",
    "#test\n",
    "p = initialiseP(seqs, w, nts)\n",
    "print (p)\n",
    "print(nts)\n",
    "print(np.random.randint(0,4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "rMqgrzYRBmjS"
   },
   "source": [
    "3\\. Implémenter une fonction `initialiseZ` pour initialiser la matrice $Z$ à uns. Rappelez-vous que la dimension de $Z$ est $nbSeq \\times (lenSeq -w +1)$, où $nbSeq$ est le nombre de sequences et $lenSeq$ est la taille/longueur des sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "id": "mghUiiW4BmjT"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 2, 1, 2]\n",
      "[[0. 1. 0. 0.]\n",
      " [0. 0. 1. 0.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 1. 0.]]\n"
     ]
    }
   ],
   "source": [
    "# \n",
    "\n",
    "def initialiseZ(seqs, w):\n",
    "    \"\"\"\n",
    "    Initialise Z\n",
    "    input seqs : list of sequences\n",
    "    input w : motif length\n",
    "    output Z :  matrix of motif start positions\n",
    "    \"\"\"\n",
    "    \n",
    "#     Z = []\n",
    "    Z = np.zeros((len(seqs), len(seqs[0])-w+1))\n",
    "    l=[]\n",
    "    for i in range(len(seqs)):\n",
    "        r=np.random.randint(0,len(seqs[0])-w+1)\n",
    "        l.append(r)\n",
    "    print(l)\n",
    "    \n",
    "    for i in range(len(Z)):\n",
    "        Z[i][l[i]]=1\n",
    "   \n",
    "    return Z\n",
    "\n",
    "Z = initialiseZ(seqs, w)\n",
    "print(Z)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "fbSoL4ywBmjU"
   },
   "source": [
    "4\\. Écrivez une fonction `E_step` pour le pas Expectation qui estime $Z$ à partir de  $p_{c,k}$. \n",
    "Écrivez aussi une fonction `normaliseZ` pour normaliser $Z$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "id": "9oa3r0j3BmjV"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 3, 0, 2]\n",
      "Z  : \n",
      "\n",
      "[[1. 0. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [1. 0. 0. 0.]\n",
      " [0. 0. 1. 0.]]\n",
      "P  : \n",
      "\n",
      "[[0.25       0.5        0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.5        0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.5       ]]\n",
      "i =  0\n",
      "j =  0\n",
      " z i j  1.0\n",
      " P i  [0.25       0.5        0.16666667 0.16666667]\n",
      " seq[i] [j] lettre G\n",
      "i =  1\n",
      "j =  3\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.16666667]\n",
      " seq[i] [j] lettre A\n",
      "i =  2\n",
      "j =  0\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.5        0.16666667]\n",
      " seq[i] [j] lettre A\n",
      "i =  3\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.5       ]\n",
      " seq[i] [j] lettre A\n",
      " LISTE IND [2, 0, 0, 0]\n",
      "seqs[i][0:listeind[i]] GT\n",
      "prod 0.0625\n",
      "minind 1\n",
      "minind 0\n",
      "minind 2\n",
      "mini [1, 0, 2]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 0\n",
      "minind 2\n",
      "mini [2, 0, 2]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 0\n",
      "minind 1\n",
      "minind 2\n",
      "mini [0, 1, 2]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 1\n",
      "minind 1\n",
      "minind 0\n",
      "mini [1, 1, 0]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.16666666666666666]\n",
      "LISTE PROD  [0.0009765625, 0.0023148148148148147, 0.0023148148148148147, 0.0023148148148148147]\n",
      "[[1. 0. 0. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [1. 0. 0. 0.]\n",
      " [0. 0. 1. 0.]]\n"
     ]
    }
   ],
   "source": [
    "def E_step(seqs, P, Z, w, alph):\n",
    "    \"\"\"\n",
    "    Implement Expectation step\n",
    "    input seqs : list of sequences\n",
    "    input P : position probability matrix\n",
    "    input Z :  matrix of motif start positions\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output Z :  matrix of motif start positions\n",
    "    \"\"\"\n",
    "    \n",
    "    Z = initialiseZ(seqs, w)\n",
    "    print(\"Z  : \\n\")\n",
    "    print(Z)\n",
    "    print(\"P  : \\n\")\n",
    "    print(P)\n",
    "    listeind=[]\n",
    "    for i in range(len(seqs)):\n",
    "        for j in range(len(seqs[0])-w+1):\n",
    "            if(Z[i][j]!=0):\n",
    "                print(\"i = \",i)\n",
    "                print(\"j = \",j)\n",
    "                print(\" z i j \",Z[i][j])\n",
    "                print(\" P i \",P[i])\n",
    "                print(\" seq[i] [j] lettre\",seqs[i][j])\n",
    "                indice=alph.index(seqs[i][j])\n",
    "                listeind.append(indice)\n",
    "    print(\" LISTE IND\",listeind)\n",
    "    listeprod=[]\n",
    "    \n",
    "    for i in range(len(seqs)):\n",
    "        prod=1\n",
    "        print(\"seqs[i][0:listeind[i]]\",seqs[i][0:listeind[i]])\n",
    "        prod=prod*(0.25**len(seqs[i][0:listeind[i]]))\n",
    "        print(\"prod\",prod)\n",
    "        mini=[]\n",
    "        pp=[]\n",
    "        for l in range(listeind[i],listeind[i]+w):\n",
    "            minind=alph.index(seqs[i][l])\n",
    "            print(\"minind\",minind)\n",
    "            mini.append(minind)\n",
    "        print(\"mini\",mini)\n",
    "        for a in mini:\n",
    "            p=P[a][i]\n",
    "            pp.append(p)\n",
    "            \n",
    "        print(\"pp\",pp)\n",
    "        for b in pp:\n",
    "            prod=prod*b\n",
    "        little=[]\n",
    "        foi=[]\n",
    "        for c in range(listeind[i]+w+1,len(seqs[0])-1):\n",
    "            ll=alph.index(seqs[i][c])\n",
    "            little.append(ll)\n",
    "        for d in little:\n",
    "            f=P[d][i]\n",
    "            foi.append(f)\n",
    "        for e in foi:\n",
    "            prod=prod*e\n",
    "            \n",
    "            \n",
    "        listeprod.append(prod)\n",
    "        \n",
    "    print(\"LISTE PROD \",listeprod)\n",
    "    \n",
    "            \n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind2=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "        \n",
    "            \n",
    "                \n",
    "                \n",
    "            \n",
    "    \n",
    "    return Z\n",
    "\n",
    "def normaliseZ(z):\n",
    "    \"\"\"\n",
    "    Normalise Z matrix\n",
    "    input Z : unnormalised matrix\n",
    "    output Zn : normalised matrix\n",
    "    \"\"\"\n",
    "    return np.array([row/sum(row) for row in z])\n",
    "\n",
    "Z = E_step(seqs, p, Z, w, nts)\n",
    "z_norm = normaliseZ(Z)\n",
    "print(z_norm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 2, 1, 2]\n",
      "Z  : \n",
      "\n",
      "[[0. 1. 0. 0.]\n",
      " [0. 0. 1. 0.]\n",
      " [0. 1. 0. 0.]\n",
      " [0. 0. 1. 0.]]\n",
      "P  : \n",
      "\n",
      "[[0.25       0.5        0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.5        0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.5       ]]\n",
      "i =  0\n",
      "j =  1\n",
      " z i j  1.0\n",
      " P i  [0.25       0.5        0.16666667 0.16666667]\n",
      " seq[i] [j] lettre T\n",
      "i =  1\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.16666667]\n",
      " seq[i] [j] lettre G\n",
      "i =  2\n",
      "j =  1\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.5        0.16666667]\n",
      " seq[i] [j] lettre C\n",
      "i =  3\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.5       ]\n",
      " seq[i] [j] lettre A\n",
      " LISTE IND [3, 2, 1, 0]\n",
      "LIND  [0, 1, 2, 3]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 0\n",
      "minind 2\n",
      "minind 0\n",
      "mini [0, 2, 0]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 0\n",
      "minind 2\n",
      "minind 0\n",
      "mini [0, 2, 0]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 0\n",
      "minind 2\n",
      "minind 0\n",
      "mini [0, 2, 0]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 0\n",
      "minind 2\n",
      "minind 0\n",
      "mini [0, 2, 0]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] AC\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 2\n",
      "minind 0\n",
      "mini [2, 2, 0]\n",
      "pp [0.5, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] AC\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 2\n",
      "minind 0\n",
      "mini [2, 2, 0]\n",
      "pp [0.5, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] AC\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 2\n",
      "minind 0\n",
      "mini [2, 2, 0]\n",
      "pp [0.5, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] AC\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 2\n",
      "minind 0\n",
      "mini [2, 2, 0]\n",
      "pp [0.5, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] CCA\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] CCA\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] CCA\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] CCA\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "LISTE PROD  [[0.00390625, 0.00390625, 0.00390625, 0.00390625], [0.010416666666666666, 0.010416666666666666, 0.010416666666666666, 0.010416666666666666], [0.0026041666666666665, 0.0026041666666666665, 0.0026041666666666665, 0.0026041666666666665], [0.00021701388888888888, 0.00021701388888888888, 0.00021701388888888888, 0.00021701388888888888]]\n",
      " Z \n",
      "[[0.00390625 0.00390625 0.00390625 0.00390625]\n",
      " [0.01041667 0.01041667 0.01041667 0.01041667]\n",
      " [0.00260417 0.00260417 0.00260417 0.00260417]\n",
      " [0.00021701 0.00021701 0.00021701 0.00021701]]\n",
      "[[0.25 0.25 0.25 0.25]\n",
      " [0.25 0.25 0.25 0.25]\n",
      " [0.25 0.25 0.25 0.25]\n",
      " [0.25 0.25 0.25 0.25]]\n"
     ]
    }
   ],
   "source": [
    "def E_step(seqs, P, Z, w, alph):\n",
    "    \"\"\"\n",
    "    Implement Expectation step\n",
    "    input seqs : list of sequences\n",
    "    input P : position probability matrix\n",
    "    input Z :  matrix of motif start positions\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output Z :  matrix of motif start positions\n",
    "    \"\"\"\n",
    "    \n",
    "    Z = initialiseZ(seqs, w)\n",
    "    print(\"Z  : \\n\")\n",
    "    print(Z)\n",
    "    print(\"P  : \\n\")\n",
    "    print(P)\n",
    "    listeind=[]\n",
    "    for i in range(len(seqs)):\n",
    "        for j in range(len(seqs[0])-w+1):\n",
    "            if(Z[i][j]!=0):\n",
    "                print(\"i = \",i)\n",
    "                print(\"j = \",j)\n",
    "                print(\" z i j \",Z[i][j])\n",
    "                print(\" P i \",P[i])\n",
    "                print(\" seq[i] [j] lettre\",seqs[i][j])\n",
    "                indice=alph.index(seqs[i][j])\n",
    "                listeind.append(indice)\n",
    "    print(\" LISTE IND\",listeind)\n",
    "    \n",
    "    lind=[]\n",
    "    for a in range(0,len(seqs[0])-w+1):\n",
    "        lind.append(a)\n",
    "        \n",
    "    print(\"LIND \",lind)\n",
    "    listelisteprod=[]\n",
    "    \n",
    "    \n",
    "    for i in range(len(seqs)):\n",
    "        listeprod=[]\n",
    "        for j in range(len(seqs[0])-w+1):\n",
    "            prod=1\n",
    "            print(\"seqs[i][0:listeind[i]]\",seqs[i][0:lind[i]])\n",
    "            prod=prod*(0.25**len(seqs[i][0:lind[i]]))\n",
    "            print(\"prod\",prod)\n",
    "            mini=[]\n",
    "            pp=[]\n",
    "            for l in range(lind[i],lind[i]+w):\n",
    "                minind=alph.index(seqs[i][l])\n",
    "                print(\"minind\",minind)\n",
    "                mini.append(minind)\n",
    "            print(\"mini\",mini)\n",
    "            for a in mini:\n",
    "                p=P[a][i]\n",
    "                pp.append(p)\n",
    "\n",
    "            print(\"pp\",pp)\n",
    "            for b in pp:\n",
    "                prod=prod*b\n",
    "            little=[]\n",
    "            foi=[]\n",
    "            for c in range(lind[i]+w+1,len(seqs[0])-1):\n",
    "                ll=alph.index(seqs[i][c])\n",
    "                little.append(ll)\n",
    "            for d in little:\n",
    "                f=P[d][i]\n",
    "                foi.append(f)\n",
    "            for e in foi:\n",
    "                prod=prod*e\n",
    "\n",
    "\n",
    "            listeprod.append(prod)\n",
    "        listelisteprod.append(listeprod)\n",
    "\n",
    "    print(\"LISTE PROD \",listelisteprod)\n",
    "    \n",
    "            \n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind2=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "        \n",
    "            \n",
    "                \n",
    "                \n",
    "    Z=np.asarray(listelisteprod)     \n",
    "    \n",
    "    print(\" Z \")\n",
    "    print(Z)\n",
    "    return Z\n",
    "\n",
    "def normaliseZ(z):\n",
    "    \"\"\"\n",
    "    Normalise Z matrix\n",
    "    input Z : unnormalised matrix\n",
    "    output Zn : normalised matrix\n",
    "    \"\"\"\n",
    "    return np.array([row/sum(row) for row in z])\n",
    "\n",
    "Z = E_step(seqs, p, Z, w, nts)\n",
    "z_norm = normaliseZ(Z)\n",
    "print(z_norm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 2, 3, 2]\n",
      "Z  : \n",
      "\n",
      "[[0. 0. 1. 0.]\n",
      " [0. 0. 1. 0.]\n",
      " [0. 0. 0. 1.]\n",
      " [0. 0. 1. 0.]]\n",
      "P  : \n",
      "\n",
      "[[0.25       0.5        0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.16666667]\n",
      " [0.25       0.16666667 0.5        0.16666667]\n",
      " [0.25       0.16666667 0.16666667 0.5       ]]\n",
      "i =  0\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.5        0.16666667 0.16666667]\n",
      " seq[i] [j] lettre C\n",
      "i =  1\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.16666667]\n",
      " seq[i] [j] lettre G\n",
      "i =  2\n",
      "j =  3\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.5        0.16666667]\n",
      " seq[i] [j] lettre G\n",
      "i =  3\n",
      "j =  2\n",
      " z i j  1.0\n",
      " P i  [0.25       0.16666667 0.16666667 0.5       ]\n",
      " seq[i] [j] lettre A\n",
      " LISTE IND [1, 2, 2, 0]\n",
      "LIND  [0, 1, 2, 3]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 3\n",
      "minind 1\n",
      "minind 0\n",
      "mini [3, 1, 0]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] GT\n",
      "prod 0.0625\n",
      "minind 1\n",
      "minind 0\n",
      "minind 2\n",
      "mini [1, 0, 2]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] GTC\n",
      "prod 0.015625\n",
      "minind 0\n",
      "minind 2\n",
      "minind 2\n",
      "mini [0, 2, 2]\n",
      "pp [0.25, 0.25, 0.25]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 2\n",
      "minind 0\n",
      "minind 2\n",
      "mini [2, 0, 2]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] G\n",
      "prod 0.25\n",
      "minind 0\n",
      "minind 2\n",
      "minind 0\n",
      "mini [0, 2, 0]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] GA\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 0\n",
      "minind 2\n",
      "mini [2, 0, 2]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] GAG\n",
      "prod 0.015625\n",
      "minind 0\n",
      "minind 2\n",
      "minind 3\n",
      "mini [0, 2, 3]\n",
      "pp [0.5, 0.16666666666666666, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 0\n",
      "minind 1\n",
      "minind 2\n",
      "mini [0, 1, 2]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] A\n",
      "prod 0.25\n",
      "minind 1\n",
      "minind 2\n",
      "minind 2\n",
      "mini [1, 2, 2]\n",
      "pp [0.16666666666666666, 0.5, 0.5]\n",
      "seqs[i][0:listeind[i]] AC\n",
      "prod 0.0625\n",
      "minind 2\n",
      "minind 2\n",
      "minind 0\n",
      "mini [2, 2, 0]\n",
      "pp [0.5, 0.5, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] ACG\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 0\n",
      "minind 2\n",
      "mini [2, 0, 2]\n",
      "pp [0.5, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] \n",
      "prod 1.0\n",
      "minind 1\n",
      "minind 1\n",
      "minind 0\n",
      "mini [1, 1, 0]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] C\n",
      "prod 0.25\n",
      "minind 1\n",
      "minind 0\n",
      "minind 2\n",
      "mini [1, 0, 2]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.16666666666666666]\n",
      "seqs[i][0:listeind[i]] CC\n",
      "prod 0.0625\n",
      "minind 0\n",
      "minind 2\n",
      "minind 3\n",
      "mini [0, 2, 3]\n",
      "pp [0.16666666666666666, 0.16666666666666666, 0.5]\n",
      "seqs[i][0:listeind[i]] CCA\n",
      "prod 0.015625\n",
      "minind 2\n",
      "minind 3\n",
      "minind 1\n",
      "mini [2, 3, 1]\n",
      "pp [0.16666666666666666, 0.5, 0.16666666666666666]\n",
      "LISTE PROD  [[0.00390625, 0.00390625, 0.0009765625, 0.000244140625], [0.0023148148148148147, 0.010416666666666666, 0.0008680555555555555, 0.00021701388888888888], [0.0023148148148148147, 0.010416666666666666, 0.0026041666666666665, 0.0006510416666666666], [0.0023148148148148147, 0.0011574074074074073, 0.0008680555555555555, 0.00021701388888888888]]\n",
      " Z \n",
      "[[0.00390625 0.00390625 0.00097656 0.00024414]\n",
      " [0.00231481 0.01041667 0.00086806 0.00021701]\n",
      " [0.00231481 0.01041667 0.00260417 0.00065104]\n",
      " [0.00231481 0.00115741 0.00086806 0.00021701]]\n",
      "[[0.43243243 0.43243243 0.10810811 0.02702703]\n",
      " [0.16753927 0.7539267  0.06282723 0.01570681]\n",
      " [0.14479638 0.65158371 0.16289593 0.04072398]\n",
      " [0.50793651 0.25396825 0.19047619 0.04761905]]\n"
     ]
    }
   ],
   "source": [
    "def E_step(seqs, P, Z, w, alph):\n",
    "    \"\"\"\n",
    "    Implement Expectation step\n",
    "    input seqs : list of sequences\n",
    "    input P : position probability matrix\n",
    "    input Z :  matrix of motif start positions\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output Z :  matrix of motif start positions\n",
    "    \"\"\"\n",
    "    \n",
    "    Z = initialiseZ(seqs, w)\n",
    "    print(\"Z  : \\n\")\n",
    "    print(Z)\n",
    "    print(\"P  : \\n\")\n",
    "    print(P)\n",
    "    listeind=[]\n",
    "    for i in range(len(seqs)):\n",
    "        for j in range(len(seqs[0])-w+1):\n",
    "            if(Z[i][j]!=0):\n",
    "                print(\"i = \",i)\n",
    "                print(\"j = \",j)\n",
    "                print(\" z i j \",Z[i][j])\n",
    "                print(\" P i \",P[i])\n",
    "                print(\" seq[i] [j] lettre\",seqs[i][j])\n",
    "                indice=alph.index(seqs[i][j])\n",
    "                listeind.append(indice)\n",
    "    print(\" LISTE IND\",listeind)\n",
    "    \n",
    "    lind=[]\n",
    "    for a in range(0,len(seqs[0])-w+1):\n",
    "        lind.append(a)\n",
    "        \n",
    "    print(\"LIND \",lind)\n",
    "    listelisteprod=[]\n",
    "    \n",
    "    \n",
    "    for e in seqs:\n",
    "        listeprod=[]\n",
    "        for i in lind:\n",
    "            prod=1\n",
    "            print(\"seqs[i][0:listeind[i]]\",e[0:i])\n",
    "            prod=prod*(0.25**len(e[0:i]))\n",
    "            print(\"prod\",prod)\n",
    "            mini=[]\n",
    "            pp=[]\n",
    "            for l in range(i,i+w):\n",
    "                minind=alph.index(e[l])\n",
    "                print(\"minind\",minind)\n",
    "                mini.append(minind)\n",
    "            print(\"mini\",mini)\n",
    "            for a in mini:\n",
    "                p=P[a][seqs.index(e)]\n",
    "                pp.append(p)\n",
    "\n",
    "            print(\"pp\",pp)\n",
    "            for b in pp:\n",
    "                prod=prod*b\n",
    "            little=[]\n",
    "            foi=[]\n",
    "            for c in range(i+w+1,len(seqs[0])-1):\n",
    "                ll=alph.index(e[c])\n",
    "                little.append(ll)\n",
    "            for d in little:\n",
    "                f=P[d][seqs.index(e)]\n",
    "                foi.append(f)\n",
    "            for f in foi:\n",
    "                prod=prod*f\n",
    "\n",
    "\n",
    "            listeprod.append(prod)\n",
    "        listelisteprod.append(listeprod)\n",
    "\n",
    "    print(\"LISTE PROD \",listelisteprod)\n",
    "    \n",
    "            \n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind2=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "#         ind1=alpha.index(seqs[i][listeind[i]])\n",
    "#         p1=P[i][ind1]\n",
    "        \n",
    "            \n",
    "                \n",
    "                \n",
    "    Z=np.asarray(listelisteprod)     \n",
    "    \n",
    "    print(\" Z \")\n",
    "    print(Z)\n",
    "    return Z\n",
    "\n",
    "def normaliseZ(z):\n",
    "    \"\"\"\n",
    "    Normalise Z matrix\n",
    "    input Z : unnormalised matrix\n",
    "    output Zn : normalised matrix\n",
    "    \"\"\"\n",
    "    return np.array([row/sum(row) for row in z])\n",
    "\n",
    "Z = E_step(seqs, p, Z, w, nts)\n",
    "z_norm = normaliseZ(Z)\n",
    "print(z_norm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "odh27oHwBmjV"
   },
   "source": [
    "5\\. Implémentez une fonction `M_step` pour le pas Maximisation qui estime $p_{c,k}$ à partir de $Z$. \n",
    "Utilisez les \"pseudocounts\" pour éviter les probabilités ègales à zero."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "id": "VleAgdFtBmjW",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[ 6.  5. 10.  3.]\n",
      "{'GTCAGG': ['GTC', 'TCA', 'CAG', 'AGG'], 'GAGAGT': ['GAG', 'AGA', 'GAG', 'AGT'], 'ACGGAG': ['ACG', 'CGG', 'GGA', 'GAG'], 'CCAGTC': ['CCA', 'CAG', 'AGT', 'GTC']}\n",
      "dicopos  {('G', 0): [('GTC', 'GTCAGG'), ('GAG', 'GAGAGT'), ('GAG', 'GAGAGT'), ('GGA', 'ACGGAG'), ('GAG', 'ACGGAG'), ('GTC', 'CCAGTC')], ('T', 1): [('GTC', 'GTCAGG'), ('GTC', 'CCAGTC')], ('C', 2): [('GTC', 'GTCAGG'), ('GTC', 'CCAGTC')], ('T', 0): [('TCA', 'GTCAGG')], ('C', 1): [('TCA', 'GTCAGG'), ('ACG', 'ACGGAG'), ('CCA', 'CCAGTC')], ('A', 2): [('TCA', 'GTCAGG'), ('AGA', 'GAGAGT'), ('GGA', 'ACGGAG'), ('CCA', 'CCAGTC')], ('C', 0): [('CAG', 'GTCAGG'), ('CGG', 'ACGGAG'), ('CCA', 'CCAGTC'), ('CAG', 'CCAGTC')], ('A', 1): [('CAG', 'GTCAGG'), ('GAG', 'GAGAGT'), ('GAG', 'GAGAGT'), ('GAG', 'ACGGAG'), ('CAG', 'CCAGTC')], ('G', 2): [('CAG', 'GTCAGG'), ('AGG', 'GTCAGG'), ('GAG', 'GAGAGT'), ('GAG', 'GAGAGT'), ('ACG', 'ACGGAG'), ('CGG', 'ACGGAG'), ('GAG', 'ACGGAG'), ('CAG', 'CCAGTC')], ('A', 0): [('AGG', 'GTCAGG'), ('AGA', 'GAGAGT'), ('AGT', 'GAGAGT'), ('ACG', 'ACGGAG'), ('AGT', 'CCAGTC')], ('G', 1): [('AGG', 'GTCAGG'), ('AGA', 'GAGAGT'), ('AGT', 'GAGAGT'), ('CGG', 'ACGGAG'), ('GGA', 'ACGGAG'), ('AGT', 'CCAGTC')], ('T', 2): [('AGT', 'GAGAGT'), ('AGT', 'CCAGTC')]}\n",
      "DICO P UP {('G', 0): 0.7500306040496986, ('T', 1): 0.07464607464607464, ('C', 2): 0.6864006864006864, ('T', 0): 0.02702702702702703, ('C', 1): 1.3379843968079261, ('A', 2): 1.2527045874751448, ('C', 0): 1.59195265077618, ('A', 1): 1.4202438544908516, ('G', 2): 1.0210346397596166, ('A', 0): 1.4202438544908516, ('G', 1): 0.7500306040496986, ('T', 2): 0.06332585390177013}\n",
      "DICOFINAL {('G', 0): 0.1875076510124247, ('T', 1): 0.018661518661518665, ('C', 2): 0.17160017160017163, ('T', 0): 0.006756756756756759, ('C', 1): 0.3344960992019816, ('A', 2): 0.31317614686878625, ('C', 0): 0.39798816269404513, ('A', 1): 0.35506096362271294, ('G', 2): 0.2552586599399042, ('A', 0): 0.35506096362271294, ('G', 1): 0.1875076510124247, ('T', 2): 0.015831463475442537}\n",
      "3\n",
      "SEQ CCAGTC\n",
      "motif CAG\n",
      "MOTIF[i] C\n",
      "indice dans boucle 1\n",
      "MOTIF[i] A\n",
      "indice dans boucle 0\n",
      "MOTIF[i] G\n",
      "indice dans boucle 2\n",
      "G\n",
      "[[0.25       0.35506096 0.35506096 0.31317615]\n",
      " [0.25       0.39798816 0.3344961  0.17160017]\n",
      " [0.25       0.18750765 0.18750765 0.25525866]\n",
      " [0.25       0.00675676 0.01866152 0.01583146]]\n"
     ]
    }
   ],
   "source": [
    "def totalNumberofCH(seqs,alph):\n",
    "\n",
    "    q = len(alph)\n",
    "    totalN = np.zeros((q))\n",
    "    for i in range(q):\n",
    "        e = alph[i]\n",
    "        for seq in seqs:\n",
    "            for s in seq:\n",
    "                if(s == e):\n",
    "                    totalN[i]+=1\n",
    "    return totalN\n",
    "                    \n",
    "def M_step(seqs, Z, w, alph):\n",
    "    \"\"\"\n",
    "    Implement Expectation step\n",
    "    input seqs : list of sequences\n",
    "    input Z :  matrix of motif start positions\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output P : position probability matrix\n",
    "    \"\"\"\n",
    "    dico={}\n",
    "    \n",
    "    for seq in seqs:\n",
    "        for i in range(len(seq)-w+1):\n",
    "            if seq not in dico:\n",
    "                dico[seq]=[seq[i:i+w]]\n",
    "            else:\n",
    "                dico[seq].append(seq[i:i+w])\n",
    "\n",
    "    P = []\n",
    "    print(dico)\n",
    "    dicopos={}\n",
    "    \n",
    "    for key,val in dico.items():\n",
    "        for v in val:\n",
    "            for i in range(len(v)):\n",
    "                if v[i]=='A':\n",
    "                    nt='A'\n",
    "                if v[i]=='C':\n",
    "                    nt='C'\n",
    "                if v[i]=='G':\n",
    "                    nt='G'\n",
    "                if v[i]=='T':\n",
    "                    nt='T'\n",
    "\n",
    "                if(v[i],i) not in dicopos:\n",
    "                    dicopos[(v[i],i)]=[(v,key)]\n",
    "                else:\n",
    "                    dicopos[(v[i],i)].append((v,key))\n",
    "       \n",
    "    print(\"dicopos \",dicopos)\n",
    "    dicoPup={}\n",
    "    for key,val in dicopos.items():\n",
    "        nt,pos=key\n",
    "        for v in val:\n",
    "            m,seq=v\n",
    "            idnt=alph.index(nt)\n",
    "            iseq=seqs.index(seq)\n",
    "            if key not in dicoPup:\n",
    "                dicoPup[key]=Z[iseq][idnt]\n",
    "            else:\n",
    "                dicoPup[key]+=Z[iseq][idnt]\n",
    "                \n",
    "    print(\"DICO P UP\",dicoPup)\n",
    "    \n",
    "    som=0\n",
    "    for i in range(len(seqs)):\n",
    "        for j in range(len(seqs[0])-w+1):\n",
    "            som+=Z[i][j]\n",
    "    \n",
    "    dicoFinal={}\n",
    "    for key,val in dicoPup.items():\n",
    "        dicoFinal[key]=(val/som)\n",
    "        \n",
    "    print(\"DICOFINAL\",dicoFinal)\n",
    "    P=initialiseP(seqs, w, nts)\n",
    "    for k,v in dicoFinal.items():\n",
    "        nt,pos=k\n",
    "        idnt=alph.index(nt)\n",
    "        P[idnt][pos+1]=v\n",
    "    return P\n",
    "totalN=totalNumberofCH(seqs,nts)\n",
    "print(totalN)\n",
    "P = M_step(seqs, z_norm, w, nts)\n",
    "print(P)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "37ciA9qLBmjY"
   },
   "source": [
    "6\\. Écrivez une fonction `likelihood` qui calcule la log-vraisemblance de l'ensemble des sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "jUMUQUwZBmjY",
    "outputId": "b679b7aa-6b80-496d-d6be-c6342913321d"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-5.545177444479562\n"
     ]
    }
   ],
   "source": [
    "def likelihood(seqs, Z, P, w, alph):\n",
    "    \"\"\"\n",
    "    Implement log likelihood function of P\n",
    "    input seqs : list of sequences\n",
    "    input Z :  matrix of motif start positions\n",
    "    input p : position probability matrix\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    output lLikelihood : log likelihood of P \n",
    "\n",
    "    \"\"\"\n",
    "    M = len(seqs)\n",
    "    L = len(seqs[0])\n",
    "    lLikelihood = np.log((L-w+1)**-M)\n",
    "    \n",
    "      \n",
    "    return lLikelihood\n",
    "\n",
    "logvraisemblance = likelihood(seqs, Z, P, w, nts)\n",
    "print(logvraisemblance)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "LFOgaK9eBmjZ"
   },
   "source": [
    "7\\. Implémentez l'algorithme Expectation-Maximisation. Vous calculerez la valeur de la log-vraisemblance totale du modèle à chaque iteration et l'algorithme prendra fin lorsque $\\Delta \\log \\text{Pr}(D | \\Theta) < \\varepsilon$. Utilisez $\\varepsilon = 1e-4$. Votre implementation devra renvoyer les paramètres du modele ($p$ et la log-likelihood associé), ainsi bien que la liste des meilleures positions du motif dans chaque sequence (matrice $Z$). Faites attention à utiliser $Z$ non-normalisé afin de trouver la log-vraisemblance!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "-ZXGq1_3Bmja"
   },
   "outputs": [],
   "source": [
    "def ExpectationMaximisation(seqs, w, alph, eps):\n",
    "    \"\"\"\n",
    "    Implement Expectation Maximisation algorithm\n",
    "    input seqs : list of sequences\n",
    "    input w : motif length\n",
    "    input alph : alphabet (nucleotides or amino acids)\n",
    "    input eps : threahold \n",
    "    output P : position probability matrix\n",
    "    output Z :  matrix of motif start positions\n",
    "    output lLikelihood : log likelihood of P \n",
    "    output pos_motif : positions of motifs in seqs\n",
    "    \"\"\"\n",
    "    P = initialiseP(seqs, w, alph)\n",
    "    Z = initialiseZ(seqs, w)\n",
    "    lLikelihood = []\n",
    "    pos_motif = []\n",
    "    \n",
    "\n",
    "    return (P, lLikelihood, Z, pos_motif)\n",
    "\n",
    "def analyse_results(results, seqs):\n",
    "    print('Les paramètres du modele:')\n",
    "    print('p:\\n',results[0])\n",
    "    print('\\n')\n",
    "    print('log-vraisemblance associé:',results[1][-1])\n",
    "    print('\\n')\n",
    "    print('Z:\\n',results[2])\n",
    "    print('\\nLes motifs:')\n",
    "    for i in range(len(seqs)):\n",
    "        print('seq'+str(i+1)+': '+ str(results[3][i]+1) + '  ' + seqs[i][results[3][i]:results[3][i]+w])\n",
    "\n",
    "    plt.plot(results[1])\n",
    "    plt.xlabel('Iteration')\n",
    "    plt.ylabel('Log-vraisemblance')\n",
    "    plt.show()\n",
    "\n",
    "eps = 10**-4\n",
    "EMResults = ExpectationMaximisation(seqs, w, nts, eps)\n",
    "analyse_results(EMResults,seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MZsOsS8yBmjb"
   },
   "source": [
    "8\\. Qu'est-ce que vous observez en exécutant l'algorithme EM plusieurs fois? Justifiez votre réponse."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HgXXj3dcBmjb"
   },
   "source": [
    "Reponse:\n",
    "\n",
    "<font color=\"blue\">\n",
    "\n",
    "</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "WThFDcG5Bmjb"
   },
   "source": [
    "9\\. Pour éviter le problème identifié au point précedent, écrivez une fonction `EM_iteratif` qui exécute l'algorithme `EM` $N$ fois ($N=10$) et qui prend les paramètres associés à la meilleure log-vraisemblance. Trouvez-vous les bons motifs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "hiMm-jSkBmjc"
   },
   "outputs": [],
   "source": [
    "def EM_iteratif(N, seqs, w, alph, eps):\n",
    "    \"\"\"\n",
    "    Implement a iterative version of Expectation Maximisation algorithm\n",
    "    input N : number of iterations\n",
    "    input seqs : list of sequences\n",
    "    input w : motif length\n",
    "    input eps : threahold \n",
    "    output bestModel : the parameter of the best model\n",
    "    \"\"\"\n",
    "  \n",
    "    bestModel = P, Z, lLikelihood, pos_motif\n",
    " \n",
    "    return bestModel\n",
    "\n",
    "\n",
    "eps = 10**-4\n",
    "N=10\n",
    "meilleurEM = EM_iteratif(N, seqs, w, nts, eps)\n",
    "analyse_results(meilleurEM,seqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "us2JZ9SRBmjc"
   },
   "source": [
    "10\\. Appliquez votre algorithme `EM` à l'ensemble des séquence du fichier `trainingSequences.txt` en utilisant $w=10$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "9yc5bUMIBmjd"
   },
   "outputs": [],
   "source": [
    "w= 10\n",
    "input_f = \"trainingSequences.txt\"\n",
    "seqs_train = read_training_file(input_f)\n",
    "print (seqs_train)\n",
    "eps = 10**-4\n",
    "N=10\n",
    "meilleurEM = EM_iteratif(N, seqs_train, w, nts, eps)\n",
    "analyse_results(meilleurEM,seqs_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "r3ddFiW2Bmjd"
   },
   "source": [
    "11\\. Construire un LOGO pour le motif prédit avec le service <i>WebLogo</i>. Pour cela, identifiez le motif dans chaque séquence, utiliser <i>ClustalOmega</i> pour les aligner et puis <i>WebLogo</i> pour générer le LOGO. Ajouter le LOGO à votre réponse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "sR5YLgIZBmjd"
   },
   "outputs": [],
   "source": [
    "fhandler = open('motifs.txt','w')\n",
    "for i in range(len(seqs_train)):\n",
    "    fhandler.write('>motif'+str(i+1)+'\\n')\n",
    "    fhandler.write(seqs_train[i][meilleurEM[3][i]:meilleurEM[3][i]+w])\n",
    "    fhandler.write('\\n')\n",
    "fhandler.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0lu09z4zBmje"
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "YfmV6fjVBmje"
   },
   "source": [
    "12\\. Comparez les motifs trouvés par votre programme avec les motifs du fichier `testingSequences.txt`, où les vrais motifs sont montrés en lettre majuscule. Quelle est la performance de votre programme? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "jh36eh-vBmje"
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "motifs_extracted = []\n",
    "motifs_real = []\n",
    "\n",
    "\n",
    "input_f = \"testingSequences.txt\"\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "TME9_2022.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
